homeassistant/components/keymitt_ble/config_flow.py

```diff
@@ -0,0 +1,156 @@
+        )
+        self._discovered_adv = parsed
+        self.context["title_placeholders"] = {
+            "name": name_from_discovery(self._discovered_adv),
+        }
+        return await self.async_step_init()
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle a flow initialized by the user."""
+        # This is for backwards compatibility.
+        return await self.async_step_init(user_input)
+
+    async def async_step_init(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Check if paired."""
+        errors: dict[str, str] = {}
+
+        if discovery := self._discovered_adv:
+            self._discovered_advs[discovery.address] = discovery
+        else:
+            current_addresses = self._async_current_ids()
+            for discovery_info in async_discovered_service_info(self.hass):
+                self._ble_device = discovery_info.device
+                address = discovery_info.address
+                if (
+                    address in current_addresses
+                    or address in self._discovered_advs
+                ):
+                    continue
+                parsed = parse_advertisement_data(
+                    discovery_info.device, discovery_info.advertisement
+                )
+                if parsed:
+                    self._discovered_advs[address] = parsed
+
+        if not self._discovered_advs:
+            return self.async_abort(reason="no_unconfigured_devices")
+
+        if user_input is not None:
+            self._name = name_from_discovery(self._discovered_adv)
+            self._bdaddr = user_input[CONF_BDADDR]
+            await self.async_set_unique_id(
+                self._bdaddr, raise_on_progress=False
+            )
+            self._abort_if_unique_id_configured()
+            return await self.async_step_link()
+
+        return self.async_show_form(
+            step_id="init",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_BDADDR): vol.In(
+                        {
+                            address: f"{parsed.data['local_name']} ({address})"
+                            for address, parsed in self._discovered_advs.items()
+                        }
+                    )
+                }
+            ),
+            errors=errors
+        )
+
+    async def async_step_link(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Given a configured host, will ask the user to press the button to pair."""
+        errors: dict[str, str] = {}
+        token = randomid(32)
+        self._client = MicroBotApiClient(
+            device=self._ble_device,
+            token=token,
+        )
+        assert self._client is not None
+        if user_input is None:
+            await self._client.connect(init=True)
```

Comment:
Does this need to disconnect ?

Reply:
No the bot is far more responsive if connection is maintained. I'm disconnecting with async_will_remove_from_hass (not sure if needed but seems tidy).

Having said that, on hass restart it doesn't hold connection and disconnects itself after every command. Trying to figure that out. Works fine though as integration checks connection before sending. Just a slight delay.

Reply:
As soon as the flow completes, your library objects will get destroyed so you'll be leaving the connection open and then the setup will happen and try to establish it again but it will already be open on the stack.

Usually Ble devices don't leave leave the connection open for more than a minute or so as the adapter can only talk to a few devices at the same time.  

In switchbot we leave it open for 45 seconds or so before closing it if there are no more additional commands

Reply:
After the token initialisation the connection definitely stays open and the device maintains connection indefinitely and reacts instantaneously. Unfortunately after a HA restart it goes back to it's default behaviour of disconnecting automatically after about 10 seconds. I can't redo the initialisation here as it involves pressing a button on the bot.

The way the MicroBot app behaves I think corroborates this. One device is assigned as the local source . If a another device is closer then it switches to 'remote' mode. Seems after a restart HA is no longer the local source which is frustrating.

I guess I'll put the disconnects in for now but it's quite a sleepy device so the above behaviour is preferable.